core/shared/src/main/scala/scalaz/zio/stream/StreamChunk.scala

```diff
@@ -0,0 +1,212 @@
+
+  /**
+   * Runs the sink on the stream to produce either the sink's result or an error.
+   */
+  final def run[E1 >: E, A0, A1 >: A, B](sink: Sink[E1, A0, Chunk[A1], B]): IO[E1, B] =
+    chunks.run(sink)
+
+  /**
+   * Filters this stream by the specified predicate, retaining all elements for
+   * which the predicate evaluates to true.
+   */
+  def filter(pred: A => Boolean): StreamChunk[E, A] =
+    StreamChunk[E, A](self.chunks.map(_.filter(pred)))
+
+  def filterNot(pred: A => Boolean): StreamChunk[E, A] = filter(!pred(_))
+
+  final def foreach0[E1 >: E](f: A => IO[E1, Boolean]): IO[E1, Unit] =
+    chunks.foreach0[E1] { as =>
+      as.foldM(true) { (p, a) =>
+        if (p) f(a)
+        else IO.point(p)
+      }
+    }
+
+  final def foreach[E1 >: E](f: A => IO[E1, Unit]): IO[E1, Unit] =
+    foreach0[E1](f(_).const(true))
+
+  final def withEffect[E1 >: E, A1 >: A](f0: A1 => IO[E1, Unit]): StreamChunk[E1, A] =
+    StreamChunk(chunks.withEffect[E1, Chunk[A1]] { as =>
+      as.traverse_(f0)
+    })
+
+  def dropWhile(pred: A => Boolean): StreamChunk[E, A] =
+    StreamChunk(new Stream[E, Chunk[A]] {
+      override def fold[E1 >: E, A1 >: Chunk[A], S](s: S)(f: (S, A1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+        self
+          .foldChunks[E1, A, (Boolean, S)](true -> s) {
+            case ((true, s), as) =>
+              val remaining = as.dropWhile(pred)
+
+              if (remaining.length > 0) f(s, remaining).map(_.map(false -> _))
+              else IO.now(Step.Cont(true                                -> s))
+            case ((false, s), as) => f(s, as).map(_.map(false -> _))
+          }
+          // The cast, although redundant, is unfortunately needed to appease Scala 2.11
+          .map(_.map(_._2).asInstanceOf[Step[S]])
+    })
+
+  def takeWhile(pred: A => Boolean): StreamChunk[E, A] =
+    StreamChunk(new Stream[E, Chunk[A]] {
+      override def fold[E1 >: E, A1 >: Chunk[A], S](s: S)(f: (S, A1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+        self.foldChunks[E1, A, S](s) { (s, as) =>
+          val remaining = as.takeWhile(pred)
+
+          if (remaining.length < as.length) f(s, remaining).map(s => Step.stop(s.extract))
+          else f(s, remaining)
+        }
+    })
+
+  def zipWithIndex: StreamChunk[E, (A, Int)] =
+    StreamChunk(
+      new Stream[E, Chunk[(A, Int)]] {
+        override def fold[E1 >: E, A1 >: Chunk[(A, Int)], S](s: S)(f: (S, A1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+          chunks
+            .fold[E1, Chunk[A], (S, Int)]((s, 0)) {
+              case ((s, index), as) =>
+                val zipped = as.zipWithIndex0(index)
+
+                f(s, zipped).map(_.map(s => (s, index + as.length)))
+            }
+            // The cast, although redundant, is unfortunately needed to appease Scala 2.11
+            .map(_.map(_._1).asInstanceOf[Step[S]])
+      }
+    )
+
+  final def scan[@specialized S1, @specialized B](s1: S1)(f1: (S1, A) => (S1, B)): StreamChunk[E, B] =
```

Comment:
I think we should call these `mapAccum` or something, and make the simpler version `scan` to be more consistent with Scala collections, Akka Stream, FS2.